lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -1069,13 +1078,74 @@ SPIRVValue *SPIRVModuleImpl::addNullConstant(SPIRVType *Ty) {
 
 SPIRVValue *SPIRVModuleImpl::addCompositeConstant(
     SPIRVType *Ty, const std::vector<SPIRVValue *> &Elements) {
-  return addConstant(new SPIRVConstantComposite(this, Ty, getId(), Elements));
+  const int MaxNumElements = MaxWordCount - SPIRVConstantComposite::FixedWC;
+  const int NumElements = Elements.size();
+
+  // In case number of elements is greater than maximum WordCount and
+  // SPV_INTEL_long_constant_composite is not enabled, the error will be emitted
+  // by validate functionality of SPIRVCompositeConstant class.
+  if (NumElements <= MaxNumElements ||
+      !isAllowedToUseExtension(ExtensionID::SPV_INTEL_long_constant_composite))
+    return addConstant(new SPIRVConstantComposite(this, Ty, getId(), Elements));
+
+  auto Start = Elements.begin();
+  auto End = Start + MaxNumElements;
+  std::vector<SPIRVValue *> Slice(Start, End);
```

Comment:
Instead of making copies, I think it would be better to use `ArrayRef` here

Reply:
Already existing `addCompositeConstant` functionality receives `std::vector` as a parameter since constructor of `SPIRVConstantComposite` instruction receives `std::vector`. There is a conversion of `ArrayRef` to `std::vector` that will do copy anyway, so I'm not sure that it makes sense unless we change interface of `addCompositeConstant` and `SPIRVConstantComposite` class, and I believe such change should be done in a separate patch.

Reply:
Ok, I'm fine to do this in a separate PR